File: app/src/main/java/org/tvheadend/tvhclient/domain/data_sources/MiscData.java
package org.tvheadend.tvhclient.domain.data_sources;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.AsyncTask;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.data.db.AppRoomDatabase;
import org.tvheadend.tvhclient.domain.entity.Connection;
import org.tvheadend.tvhclient.domain.entity.ServerStatus;
import org.tvheadend.tvhclient.ui.features.settings.DatabaseClearedCallback;

import java.lang.ref.WeakReference;

import timber.log.Timber;

public class MiscData {

    private final AppRoomDatabase db;
    private static WeakReference<DatabaseClearedCallback> callback;

    public MiscData(AppRoomDatabase database) {
        this.db = database;
    }

    public void clearDatabase(Context context, DatabaseClearedCallback callback) {
        MiscData.callback = new WeakReference<>(callback);
        new ClearDatabaseTask(context, db).execute();
    }

    private static class ClearDatabaseTask extends AsyncTask<Void, Void, Void> {
        private final AppRoomDatabase db;
        private final ProgressDialog dialog;
        private final String msg;

        ClearDatabaseTask(Context context, AppRoomDatabase db) {
            this.db = db;
            this.dialog = new ProgressDialog(context);
            this.msg = context.getString(R.string.deleting_database_contents);
        }

        @Override
        protected void onPreExecute() {
            dialog.setMessage(msg);
            dialog.setIndeterminate(true);
            dialog.show();
        }

        @Override
        protected Void doInBackground(Void... voids) {
            Timber.d("Deleting database contents...");

            db.getChannelDao().deleteAll();
            db.getChannelTagDao().deleteAll();
            db.getTagAndChannelDao().deleteAll();
            db.getProgramDao().deleteAll();
            db.getRecordingDao().deleteAll();
            db.getSeriesRecordingDao().deleteAll();
            db.getTimerRecordingDao().deleteAll();
            db.getServerProfileDao().deleteAll();

            // Clear all assigned profiles
            for (Connection connection : db.getConnectionDao().loadAllConnectionsSync()) {
                connection.setLastUpdate(0);
                connection.setSyncRequired(true);
                db.getConnectionDao().update(connection);

                ServerStatus serverStatus = db.getServerStatusDao().loadServerStatusByIdSync(connection.getId());
                if (serverStatus != null) {
                    serverStatus.setHtspPlaybackServerProfileId(0);
                    serverStatus.setHttpPlaybackServerProfileId(0);
                    serverStatus.setCastingServerProfileId(0);
                    serverStatus.setRecordingServerProfileId(0);
                    db.getServerStatusDao().update(serverStatus);
                }
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            Timber.d("Deleting database contents finished");
            dialog.dismiss();
            DatabaseClearedCallback databaseClearedCallback = callback.get();
            if (databaseClearedCallback != null) {
                databaseClearedCallback.onDatabaseCleared();
            }
        }
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/domain/repository/data_source/MiscData.kt
package org.tvheadend.tvhclient.domain.repository.data_source

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.tvheadend.tvhclient.data.db.AppRoomDatabase
import org.tvheadend.tvhclient.ui.features.settings.DatabaseClearedCallback
import timber.log.Timber
import java.lang.ref.WeakReference

class MiscData(private val db: AppRoomDatabase) {

    private val ioScope = CoroutineScope(Dispatchers.IO)

    fun clearDatabase(callback: DatabaseClearedCallback) {
        MiscData.callback = WeakReference(callback)
        ioScope.launch {
            clearDatabase()
        }
    }

    private fun clearDatabase() {
        Timber.d("Deleting database contents...")

        db.channelDao.deleteAll()
        db.channelTagDao.deleteAll()
        db.tagAndChannelDao.deleteAll()
        db.programDao.deleteAll()
        db.recordingDao.deleteAll()
        db.seriesRecordingDao.deleteAll()
        db.timerRecordingDao.deleteAll()
        db.serverProfileDao.deleteAll()

        // Clear all assigned profiles
        for (connection in db.connectionDao.loadAllConnectionsSync()) {
            connection.lastUpdate = 0
            connection.isSyncRequired = true
            db.connectionDao.update(connection)

            val serverStatus = db.serverStatusDao.loadServerStatusByIdSync(connection.id)
            serverStatus.htspPlaybackServerProfileId = 0
            serverStatus.httpPlaybackServerProfileId = 0
            serverStatus.castingServerProfileId = 0
            serverStatus.recordingServerProfileId = 0
            db.serverStatusDao.update(serverStatus)
        }

        Timber.d("Deleting database contents finished")
        callback?.get()?.onDatabaseCleared()
    }

    companion object {
        private var callback: WeakReference<DatabaseClearedCallback>? = null
    }
}
